File: Python/SOLID/solid_4.py
"""
SOLID principles
4: Interface segregation principle
"""


from abc import abstractmethod


class Machine:
    """
    Interface
    """

    def print(self, document: str) -> None:
        """
        Print func
        """
        raise NotImplementedError

    def fax(self, document: str) -> None:
        """
        Fax func
        """
        raise NotImplementedError

    def scan(self, document: str) -> None:
        """
        Scan func
        """
        raise NotImplementedError


class MultiFunctionPrinter(Machine):
    """
    Implementation of multifunction printer
    """

    def print(self, document: str) -> None:
        """
        Print func
        """

    def fax(self, document: str) -> None:
        """
        Fax func
        """

    def scan(self, document: str) -> None:
        """
        Scan func
        """


class OldFashionedPrinter(Machine):
    """
    Old machine
    """

    def print(self, document: str) -> None:
        """
        Print func
        """
        # OK

    def fax(self, document: str) -> None:
        """
        Fax func
        """
        # noop

    def scan(self, document: str) -> None:
        """
        Scan func - NOT SUPPORTED
        """
        raise NotImplementedError("Printer cannot scan!")


# Better way
class Printer:
    """
    Printer class
    """

    @abstractmethod
    def print(self, document: str) -> None:
        """
        Print method
        """


class Scanner:
    """
    Scanner class
    """

    @abstractmethod
    def scan(self, document: str) -> None:
        """
        Scan method
        """


class MyPrinter(Printer):
    """
    Printer only
    """

    def print(self, document: str) -> None:
        """
        Print method
        """


class MyScanner(Scanner):
    """
    Printer only
    """

    def scan(self, document: str) -> None:
        """
        Print method
        """


class Photocopier(Printer, Scanner):
    """
    Both printing and scanning
    """

    def print(self, document: str) -> None:
        """
        Print method
        """
        print(document)

    def scan(self, document: str) -> None:
        """
        Scan method
        """


# Even better
class MultiFunctionDevice(Printer, Scanner):
    """
    Interface for Multifunctional device
    """

    @abstractmethod
    def print(self, document: str) -> None:
        """
        Print method
        """

    @abstractmethod
    def scan(self, document: str) -> None:
        """
        Scan method
        """


class MultiFunctionMachine(MultiFunctionDevice):
    """
    Implementation of interface for Multifunctional device
    """

    def __init__(self, printer, scanner) -> None:
        """
        Constructor
        """
        self.printer = printer
        self.scanner = scanner

    def print(self, document: str) -> None:
        """
        Print implementation
        """
        self.printer.print(document)

    def scan(self, document: str) -> None:
        """
        Print implementation
        """
        self.scanner.scan(document)


if __name__ == "__main__":
    device = MultiFunctionMachine(MyPrinter(), MyScanner())
    device.print("123")

File: Go/SOLID/solid_4.go
// SOLID principles
// 4: Interface segregation principle

package main

type Document struct {
}

type Machine interface {
	Print(d Document)
	Fax(d Document)
	Scan(d Document)
}

type MultiFunctionPrinter struct {
}

func (m MultiFunctionPrinter) Print(d Document) { panic("not implemented") }

func (m MultiFunctionPrinter) Fax(d Document) { panic("not implemented") }

func (m MultiFunctionPrinter) Scan(d Document) { panic("not implemented") }

type OldFashionedPrinter struct{}

func (o OldFashionedPrinter) Print(d Document) {
	panic("not implemented")
}

// Deprecated:
func (o OldFashionedPrinter) Fax(d Document) {
	panic("operation not supported")
}
func (o OldFashionedPrinter) Scan(d Document) {
	panic("operation not supported")
}

// ISP taken into account
type Printer interface {
	Print(d Document)
}

type Scanner interface {
	Scan(d Document)
}

type Fax interface {
	Fax(d Document)
}

type MyPrinter struct{}

func (m MyPrinter) Print(d Document) {}

type Photocopier struct{}

func (p Photocopier) Scan(d Document) {}

func (p Photocopier) Print(d Document) {}

type MultiFunctionDevice interface {
	Printer
	Scanner
	Fax
}

type MultiFunctionMachine struct {
	printer Printer
	scanner Scanner
	fax     Fax
}

func (m MultiFunctionMachine) Print(d Document) {
	m.printer.Print(d)
}

func (m MultiFunctionMachine) Scan(d Document) {
	m.scanner.Scan(d)
}

func (m MultiFunctionMachine) Fax(d Document) {
	m.fax.Fax(d)
}

func main() {}
